pkg/operator/ceph/cluster/osd/deviceSet.go

```diff
@@ -60,12 +60,19 @@ func (c *Cluster) prepareStorageClassDeviceSets(config *provisionConfig) []rookv
 					continue
 				}
 
-				if pvcTemplate.Name == bluestorePVCData {
+				// The PVC type must be from a predefined set such as "data" and "metadata". These names must be enforced if the wal/db are specified
+				// with a separate device, but if there is a single volume template we can assume it is always the data template.
+				pvcType := pvcTemplate.Name
+				if len(storageClassDeviceSet.VolumeClaimTemplates) == 1 {
```

Comment:
I'm not sure to understand what this is fixing that is not fixed by L52 already?

Reply:
This is fixing the case where the name is not blank, but neither is it data or metadata. It is something else randomly named since the names were not enforced from v1.1 and v1.2. Line 52 just fixes the case where it is blank.

Reply:
I can chime in here for a little more context since I'm the one who submitted the report that spawned this PR.

When I originally set up Rook on my cluster, it was version 1.1. And I used this schema for mt PVC-backed storage spec:

```
      volumeClaimTemplates:
      - metadata:
          name: ceph-data
        spec:
          resources:
            requests:
              storage: 1Ti
          storageClassName: gp2
          volumeMode: Block
          accessModes:
            - ReadWriteOnce
```

Here I used `metadata.name: ceph-data`. I believe I got the name `ceph-data` from an official example _somewhere_, but I'm now thinking I must have gotten it from an unofficial source.

But in any case, prior to Rook version 1.3 the docs did not specify that the PVC template name was required to follow any particular naming schema nor did the code verify that it was following any particular naming schema.

Version 1.3 both began enforcing that the PVC template name be either `metadata` or `data` (`wal` was added to the list later). The code that began enforcing it was added in https://github.com/rook/rook/pull/5639, and the documentation explaining the naming requirement was added in https://github.com/rook/rook/commit/c08c3ced05a27eccdbb7d9b41ed982c4c8ddf048#diff-5c2b799b4c387403b6f8c339b685e071R748-R752, both changes added in version 1.3.

Unfortunately, because I already have a cluster with pvc template named `ceph-data`, this makes it impossible for me to upgrade past version 1.2, as it immediately fails with the error "no data template" because it could not find a pvc template named `data`.

This PR would fix the issue for me. I only have one pvc template defined and it is a data volume but it is neither named `data` nor left blank (the latter would be the case that L52 deals with as travis pointed out).